--- setup-egress-for-space	2025-02-07 17:06:31.209387961 -0800
+++ redeploy-egress-for-space	2025-06-02 15:16:29.905157469 -0700
@@ -25,15 +25,14 @@
 fi
 
 usage="
-$0: (Re-)deploy egress proxy for an environment
+$0: Redeploy egress proxy for an environment, leaving username:password credentials unchanged.
 
 Usage:
    $0 -h
-   $0 [--restart] [egressspace] [egressorg]
+   $0 [egressspace] [egressorg]
 
 Options:
 -h:            show help and exit
---restart:     restart apps (Apps need a restart to pick up the new proxy credentials)
 egressspace:   space with public egress (default: shared-egress)
 egressorg:     org where the public egress space is (default: <currentorg>)
 
@@ -45,11 +44,14 @@
 - You must be a SpaceDeveloper in the current space and the egress_space.
 
 NOTES:
-- This script deploys a single egress proxy app to serve all apps in the target space.
-- Files <allow|deny>.acl will be created if they don't exist
-- Any S3 buckets bound to an app will automatically be allowed
-- The username:password credentials of an existing proxy WILL BE CHANGED.
-- Run setup-egress-for-apps to reset the PROXYROUTE variable and restart the apps to use it.
+- This script redeploys a single egress proxy app for the target space.
+- It will update:
+  - The caddy binary
+  - allow.acl and deny.acl. Any S3 buckets bound to an app in the target space
+    will automatically be allowed
+- The username:password credentials the proxy WILL NOT BE CHANGED.
+- A proxy will not be launched if it is not already present! This is just for updating existing proxy apps.
+- Apps in the target space WILL NOT BE RESTARTED. That's the point of using this script -- less downtime.
 "
 
 if [[ "$1" == "-h" ]]; then
@@ -61,12 +63,6 @@
 startorg=$(   cf target | grep org:   | awk '{ print $2 }')
 startspace=$( cf target | grep space: | awk '{ print $2 }')
 
-restartApps=0
-if [[ "$1" == "--restart" ]]; then
-    restartApps=1
-    shift
-fi
-
 egressspace=${1:-shared-egress}
 egressorg=${2:-$startorg}
 
@@ -97,6 +93,14 @@
 fi
 echo "Using egress proxy app: $proxyname"
 
+# Verify $proxyname app exists:
+assertCFApp $proxyname
+
+# Get username and password values from $proxyname app env:
+PROXYENV=$(cf env "$proxyname")
+PROXY_USERNAME=$(echo "$PROXYENV" | grep PROXY_USER | awk '{print $2}')
+PROXY_PASSWORD=$(echo "$PROXYENV" | grep PROXY_PASSWORD | awk '{print $2}')
+
 popspace
 
 # If DRYRUN is set, don't run the commands, but echo them.
@@ -137,10 +141,9 @@
 $output touch $SCRIPT_DIR/apps-egress-allow.acl
 
 # Assemble vars.yml file content
-
-echo "Regenerate user/pass for use by proxy"
-username="$(uuidgen)"
-password="$(uuidgen)"
+echo "Retrieve username:password of existing proxy"
+username=$PROXY_USERNAME
+password=$PROXY_PASSWORD
 proxyroute="https://$username:$password@$proxyname.apps.internal:61443"
 
 # Auto-allow bound buckets
@@ -199,23 +202,3 @@
 fi
 
 popspace
-
-if [ $restartApps -eq 1 ]; then
-    # APPNAMES=""
-    # for app in $applist ; do
-    # 	if [ $APPNAMES ]; then
-    # 	    APPNAMES="${APPNAMES},"
-    # 	fi
-    # 	APPNAMES="${APPNAMES}${app}"
-    # done
-    # echo "$SCRIPT_DIR/setup-egress-for-apps $APPNAMES"
-    # $output $SCRIPT_DIR/setup-egress-for-apps $APPNAMES
-
-    # While it seems it would be desirable to run setup-egress-for-apps with the
-    # whole list at once, in practice that script will stop after a slow-to-restart
-    # app doesn't complete in time.
-    for app in $applist ; do
-	echo "$SCRIPT_DIR/setup-egress-for-apps $app"
-	$output $SCRIPT_DIR/setup-egress-for-apps $app
-    done
-fi
